consensus/istanbul/backend/announce.go

```diff
@@ -194,12 +143,28 @@ func (sb *Backend) announceThread() {
 				logger.Warn("Error gossiping all signed announce versions")
 			}
 
+		case <-announceRetryTimerCh: // If this is nil, this channel will never receive an event
+			announceRetryTimer = nil
+			announceRetryTimerCh = nil
+			sb.startGossipAnnounceTask()
+
 		case <-sb.generateAndGossipAnnounceCh:
 			if shouldAnnounce {
-				err := sb.generateAndGossipAnnounce()
+				// This node may have recently sent out an announce message within
+				// the gossip cooldown period imposed by other nodes.
+				// Regardless, send the announce so that it will at least be
+				// processed by this node's peers. This is helpful when a network
+				// is first starting up.
+				success, err := sb.generateAndGossipAnnounce()
 				if err != nil {
 					logger.Warn("Error in generating and gossiping announce", "err", err)
 				}
+				// If a retry hasn't been scheduled already by a previous announce,
+				// schedule one.
+				if success && announceRetryTimer != nil {
```

Comment:
Directly, I think the second condition is flipped and is intended to be `announceRetryTimer == nil`. More generally, I think this logic is problematic because we should continue to periodically announce until we obtain that peers enode, either through their announcement or a directly shared enode certificate. If we stop either when the announce is multicasted, or when an error occurs in multicasting, we may not reach them and its possible to get stuck in a state where a pair of validators are ignorant of each others enode and not announcing, which would be bad. I do like the intention here, of shutting down this timer while there are no announcements to make. We could accomplish this by checking for the condition that that the generated announce message is empty. If it is empty, then there is no announcement to be made and as long as we restart the timer when there is a new announcement to be made, this will be a good way to increase the responsiveness to changes.